source/core/GenericFunctions.cpp

```diff
@@ -1695,8 +1695,17 @@ VIREO_FUNCTION_SIGNATUREV(ArrayConcatenateInternal, ArrayConcatenateInternalPara
         IntIndex minInputRank = outputRank;
         ArrayDimensionVector tempDimensionLengths, origDimensionLengths, origSlab;
         Int32 originalOuterDimSize = pDest->DimensionLengths()[outputRank-1];
+        Boolean isDestSameAsFirstInput = false;  // true implies inplaceness case
         for (i = 0; i < numInputs; i++) {
             TypedArrayCoreRef arrayInput = *((TypedArrayCoreRef *) inputs[i]);
+            if (pDest == arrayInput) {
+                if (i == 0) {
+                    isDestSameAsFirstInput = true;
+                } else {
+                    THREAD_EXEC()->LogEvent(EventLog::kHardDataError, "Illegal ArrayConcatenate inplaceness");
```

Comment:
Moved this check early so that we bail out soon for this case where output is same as an input that is not the first input.